department-of-veterans-affairs/va.gov-cms
# VACMS-14779: jsonapi banners

## Description
Closes #14779.  Accompanies this PR in the next-build repo: https://github.com/department-of-veterans-affairs/next-build/pull/148

This PR does 3 things:
- Adds a check to prevent DB exceptions thrown when accessing the endpoint from `BannerAlerts.php`. This was consistently happening on the route for the Lovell Federal Stories Listing page.
- Trims JSON:API response data for a number of things: 
  - banner content types: `node--promo_banner`, `node--banner`, `node--full_width_banner_alert`
  - news story & person profile content types
  - media images
  - disables a lot of CMS-only resources from the base `/jsonapi` endpoint in general (FE doesn't need to know about flags, messages, CMS only blocks or content types (KB articles), etc). There are likely more things that can be disabled too, but we can do that in chunks as we identify them. All resources can easily be re-enabled if we do need them.
- Adds an initial readme for JSON:API related info.

## Testing done
I used the tugboat environment generated by the initial PR for this work https://github.com/department-of-veterans-affairs/va.gov-cms/pull/14826 (remade due to bad branch name) to verify that static pages could be built without generating errors with the banner endpoint enabled. Reducing the size of the JSON:API responses helped speed this up, and got rid of the `Unexpected < in JSON at position 0` errors when hitting the banner endpoint for a given slug.

Locally, I was able to hit https://va-gov-cms.ddev.site/jsonapi/banner-alerts?item-path=/lovell-federal-health-care/stories without generating a DB error.

## QA steps
Visit `/jsonapi/banner-alerts?item-path=/lovell-federal-health-care/stories` on the generated tugboat environment. It should resolve correctly. Check the drupal logs. There should be no exceptions raised.

You can check https://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/jsonapi/banner-alerts?item-path=/lovell-federal-health-care/stories/ to see what the response is currently. The Drupal logs will have more information related to that exception.
![Screenshot 2023-08-21 at 10 18 21 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/11279744/8598e77e-d944-42ad-81e9-8b525db10416)


To check the response size reduction:
- View differences between `/jsonapi` and the current https://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/jsonapi/
- View differences between `/jsonapi/node/news_story` and https://content-build-medc0xjkxm4jmpzxl3tfbcs7qcddsivh.ci.cms.va.gov/jsonapi/node/news_story
before:
![Screenshot 2023-08-21 at 10 23 08 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/11279744/8ffecb37-8e2b-4cf5-962a-76d1e3811bd4)

after:
![Screenshot 2023-08-21 at 10 23 03 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/11279744/472910cd-df66-4bb7-8d35-176bffea8888)


### Definition of Done


### Select Team for PR review
